clang/lib/Sema/SemaChecking.cpp

```diff
@@ -1938,11 +1938,25 @@ Sema::CheckBuiltinFunctionCall(FunctionDecl *FDecl, unsigned BuiltinID,
     if (Context.BuiltinInfo.isAuxBuiltinID(BuiltinID)) {
       assert(Context.getAuxTargetInfo() &&
              "Aux Target Builtin, but not an aux target?");
-
+      /*
+      'CheckTSBuiltinFunctionCall' function checks for an aux-target builtin and also validates it 
+      against the correct target. */
       if (CheckTSBuiltinFunctionCall(
               Context.getAuxTargetInfo()->getTriple().getArch(),
-              Context.BuiltinInfo.getAuxBuiltinID(BuiltinID), TheCall))
-        return ExprError();
+              Context.BuiltinInfo.getAuxBuiltinID(BuiltinID), TheCall)) {
+                /*At this point, we know that the builtin is an aux-builtin and also checked.
+                  AUX builtins are not allowed to appear inside device code.
+                  To handle this, we use the notion of "deferred diagnostics", where we are compiling
+                  for the device, but we don't know that this function will be codegen'ed
+                  for device yet. So we create a diagnostic which is emitted if and when we
+                  realize that the function will be codegen'ed
+                */
+                if (getLangOpts().SYCLIsDevice) {
+                    SYCLDiagIfDeviceCode(TheCall->getBeginLoc(), diag::err_aux_target_builtin_in_device_code);
+                    return ExprError();
```

Comment:
Don't return an expr-error in this branch, since this might end up being valid code (in the event that this is NOT device code).  You want to diagnose and let it continue.

Reply:
if `getLangOpts().SYCLIsDevice` returns true, doesn't it mean we are in device code?
Maybe I got that part wrong? 
If that's the case, then this if block will execute only for device code and its okay to return the error right?

Reply:
"True" means we are in device compilation, not necessarily in device  code.  It could be host code being compiled during the device compilation (which will be thrown out later).  However, we don't want to return ExprError in that case, since it defeats other checks later (when we might not be emitting an error!).

Reply:
Ok. Makes sense now.